# Bioconductor Package 
## Overview of the BioConductor package 

## Installation of BioConductor package
To install BioConductor run the below command

```
if (!requireNamespace("BiocManager", quitely=TRUE))
     install.packages("BiocManager")
 BiocManager::install("ChemmineR")
```

## Importing Data in R
Depending on the type of data file, R has a varying methods to load the data. The below list shows how to load different types of files to the R program. 

- `data <- read_csv("path/to/your/file.csv")` for csv file
- `data <- read_excel("path/to/your/file.xlsx", sheet = "Sheet1")` for excel file
- `data <- fread("path/to/your/file.csv")` for large csv files.
- `data <- read_sav("path/to/your/file.sav")` for SPSS files.
- `data <- read_dta("path/to/your/file.dta")` for stata files.
- `data <- read_sas("path/to/your/file.sas7bdat")` for SAS files.

## Writing Script of Code in R 
In R, the use can interact directly with the console or prepare a script inform of `*.R` or `*.Rmd` (R Markdown) files. 
Script are useful as they allow one to;

- easily pick up where they left off on a project and progressively build and refine code and analysis. 
- easily reproduce analysis run earlier and documentation. This also saves time on repetetive tasks. 
- do a batch processing such that tasks are automotated instead of running them one by one. 
- debug the code and find errors line by line. 
- version control such as git that tracks the changes and enables collaboration with other developers.

A script is a set of commands that usually includes commenting on what each piece of code intended to do 

## Data Interface
Follow the following steps when working with data from external interfaces; 

- Refer CSV working file(filepath) e.g `test.csv`
- Get the current working directory, `getwd()`
- Set the current working directory `setwd()`. 

### Reading the *.csv file 
In this section, we will work with two csv files, `test.csv` and `data.csv`. Both files are stored in a directory named `data`. Lets load the `data.csv` to the program. 

```{r}
person_info <- read.csv("data/data.csv")
```

The `read.csv()` function automatically converts the loaded to a data frame. `is.data.frame()` is used to find if a variable is a data frame

```{r}
# just to confirm if the said data is a data frame 
print(is.data.frame(person_info))
```
The `person_info` data set is confirmed to be a variable.
The `ncol()` function counts the number of columns/variables/features in a data frame
```{r}
print(ncol(person_info)) # count the number of columns 
```

This data frame has 6 columns. Alternatively, the `nrow()` function counts the number of rows/observations/records. 

```{r}
print(nrow(person_info)) # count the number of records/rows/observations
```

This data set has 10 rows and 6 columns. 

### Brief data analysis
When the function, `attach()` is called to a data frame, the R program can refer to specific column without associating it with the data frame name. In this section the `person_info` will be attached to R and dive shallow into data analysis. 

Calculate the mean of the salary
```{r}
# attach the file inorder to avoid the $ sign
attach(person_info)

mean(Salary)
```

The average salary in the data set is 65200. The `names()` function is called to find the column names of the data set. 
```{r}
# find the column names 
names(person_info)
```


Now, lets jump to the protein data which is stored in the `test.csv` file. The `head()` function is used to show the first 6 records of the dataset
```{r}
# Load the data set 
proteindata <- read.csv("data/test.csv")
head(proteindata)
```


Count the number of rows and columns in the protein dataset
```{r}
ncol(proteindata)
nrow(proteindata)
```

The data frame has 10 rows and 6 columns. Alternatively, the `dim()` can be used, it returns the number of rows and columns respectively. 
```{r}
dim(proteindata)
```

Get the protein detail having the most residues and the max residues value. 
```{r}
objname1 <- max(proteindata$NO_RESIDUES)
objname1

objname2 <- subset(proteindata, NO_RESIDUES == max(NO_RESIDUES))
objname2
```

The `max()` function is used to find the maximum value in a vector or column. In this code, `max(proteindata$NO_RESIDUES)` returns the highest number of residues (`NO_RESIDUES`) found in the `proteindata` dataset, and this value is stored in `objname1`.

The `subset()` function is used to filter rows in a data frame that meet certain conditions. Here, `subset(proteindata, NO_RESIDUES == max(NO_RESIDUES))` returns all rows from the `proteindata` dataset where the `NO_RESIDUES` column matches the maximum value, and the result is stored in `objname2`.

- Get the protein detail having entry in PDB database
```{r}
objname3 <- subset(proteindata, DBNAME == "PDB")
print(objname3)
```

The `proteindata` dataset is filtered to select rows where the `DBNAME` column is equal to `"PDB"`. This means it extracts all proteins that are listed in the PDB database. The resulting subset is stored in `objname3`.


Get the protein detail having more than 3 dimer from PDB database
```{r}
objname5 <- subset(proteindata, as.Date(DEPOSITED_DATE) > as.Date("2014-05-05"))
print(objname5)
```

The subset function is used to filter the `proteindata` dataset to include only rows where `DBNAME` is `"PDB"` and `DIMER_COUNT` is greater than 3.

The filtered data is stored in `objname5`

Now that we have the basic concept of how R works, lets introduce the Bioconductor package. 

## Introduction to Bioconductor
This open-source project provides tools for analyzing and understanding high-throughput genomic data. It includes a broad selection of packages for both bioinformatics and chemoinformatics. Before we explore the ChemmineR package in depth, let’s start with a few basic concepts.

### Exploring Genomic data 
The `GenomicRanges` package is used to work with genomic intervals. It installed by: 
```
BiocManager::install("GenomicRanges")
```

Lets now import the genomic ranges package to the program
```{r error=FALSE, warning=FALSE, message=FALSE}
library(GenomicRanges)
```

`GRanges` objects represent genomic ranges, allowing you to manipulate genomic intervals and align them with genomic data like gene annotations.

```{r}
# Create a simple genomic range object
gr <- GRanges(seqnames = "chr1", ranges = IRanges(start = 100, end = 200))
gr
```

### Visualizing Gene Expression Data 
The `DESeq2` package is used to find the differential gene expression analysis. It is widely used to identify genes that are differentially expressed under different conditions. Its is installed by; 
```
BiocManager::install("DESeq2")
```
Load the library
```{r error=FALSE, warning=FALSE, message=FALSE}
library(DESeq2)
```

`DESeq2` handles RNA-Seq data to model and estimate the expression levels of genes, providing statistical tools for identifying differentially expressed genes.

```{r}
# Example dataset
dds <- makeExampleDESeqDataSet()
dds <- DESeq(dds)
res <- results(dds)
plotMA(res)
```

### Annotating Genomic Data
The `AnnotationDbi` and `org.Hs.eg.db` packages are used to annotate gene identifiers with gene names, descriptions and other relevant metadata. The packages can be installed by;
```
BiocManager::install("org.Hs.eg.db")
```

Load the `org.Hs.eg.db` library.

```{r error=FALSE, warning=FALSE, message=FALSE}
library(org.Hs.eg.db)
```

The packages provide a framework for mapping and retrieving biological data from databases. It’s crucial for adding biological context to your data, such as converting gene IDs to gene names

```{r}
# Example: Map Entrez IDs to gene symbols
gene_ids <- c("1", "2", "3")
symbols <- mapIds(org.Hs.eg.db, keys = gene_ids, column = "SYMBOL", keytype = "ENTREZID")
symbols
```

### Working with Sequence Data 
The popular `BioStrings package is used to analyze the protein, RNA and DNA sequences. The library is installed as the ones before. 
```
BiocManager::install("Biostrings")
```
Load the library

```{r error=FALSE, warning=FALSE, message=FALSE}
library(Biostrings)
```

Lets manipulate  the DNA String `"ATCGGCTA"` by reverse complimenting. 

```{r}
# Create a DNAString object
dna_seq <- DNAString("ATCGGCTA")
reverseComplement(dna_seq)
```

This Bioconductor is an essential toolkit for anyone working with genomic data. Its packages are designed to facilitate a wide range of bioinformatics tasks, from basic sequence manipulation to complex statistical analyses. 